File: hackerrank/30 days of code/day 18, queues and stacks/queues_and_stacks.py
# https://www.hackerrank.com/challenges/30-queues-stacks/problem?isFullScreen=true

def main():
    word = input()
    obj = init_queue_and_stack(word)
    check_if_palindrome(obj, word)


def init_queue_and_stack(word):
    obj = QueueAndStack()
    for ch in word:
        obj.enqueue_character(ch)
        obj.push_character(ch)
    return obj


class QueueAndStack:

    def __init__(self):
        self.my_queue = []
        self.my_stack = []

    def enqueue_character(self, ch):
        self.my_queue.append(ch)

    def push_character(self, ch):
        self.my_stack.append(ch)

    def dequeue_character(self):
        return self.my_queue.pop(0)

    def pop_character(self):
        return self.my_stack.pop()


def check_if_palindrome(obj, word):
    is_palindrome = True

    for _ in range(int(len(word) / 2)):
        if obj.dequeue_character() != obj.pop_character():
            is_palindrome = False
            break

    not_word = ' ' if is_palindrome else ' not '
    print(f"The word, {word}, is{not_word}a palindrome.")


if __name__ == '__main__':
    main()

File: hackerrank/30 days of code/day 18, queues and stacks/queues_and_stacks.cpp
// https://www.hackerrank.com/challenges/30-queues-stacks/problem?isFullScreen=true

#include <iostream>
#include <queue>
#include <string_view>
#include <stack>

using namespace std;

class Queue_And_Stack {
private:
    queue<char> my_queue;
    stack<char> my_stack;

public:
    void enqueue_character(const char ch)
    {
        my_queue.push(ch);
    }

    void push_character(const char ch)
    {
        my_stack.push(ch);
    }

    char dequeue_character()
    {
        char ch{my_queue.front()};
        my_queue.pop();
        return ch;
    }

    char pop_character()
    {
        char ch{my_stack.top()};
        my_stack.pop();
        return ch;
    }
};

Queue_And_Stack init_queue_and_stack(const string_view& word);
void check_if_palindrome(Queue_And_Stack& obj, const string_view& word);

int main()
{
    string word;
    getline(cin, word);
    Queue_And_Stack obj{init_queue_and_stack(word)};
    check_if_palindrome(obj, word);
    return 0;
}

Queue_And_Stack init_queue_and_stack(const string_view& word)
{
    Queue_And_Stack obj;
    for (auto c : word) {
        obj.enqueue_character(c);
        obj.push_character(c);
    }
    return obj;
}

void check_if_palindrome(Queue_And_Stack& obj, const string_view& word)
{
    bool is_palindrome{true};

    for (auto half_length{word.size() / 2}; half_length--;)
        if (obj.dequeue_character() != obj.pop_character()) {
            is_palindrome = false;
            break;
        }

    string_view not_word{is_palindrome ? " " : " not "};
    cout << "The word, " << word << ", is" << not_word << "a palindrome.";
}
